--- a/net/minecraft/world/level/chunk/ChunkGeneratorStructureState.java
+++ b/net/minecraft/world/level/chunk/ChunkGeneratorStructureState.java
@@ -41,22 +_,51 @@
     private final Map<ConcentricRingsStructurePlacement, CompletableFuture<List<ChunkPos>>> ringPositions = new Object2ObjectArrayMap<>();
     private boolean hasGeneratedPositions;
     private final List<Holder<StructureSet>> possibleStructureSets;
+    public final io.papermc.paper.configuration.WorldConfiguration.StructureSeeds conf; // Paper - Add missing structure set seed configs
 
     public static ChunkGeneratorStructureState createForFlat(
-        final RandomState randomState, final long levelSeed, final BiomeSource biomeSource, final Stream<Holder<StructureSet>> structureOverrides
+        final RandomState randomState, final long levelSeed, final BiomeSource biomeSource, final Stream<Holder<StructureSet>> structureOverrides, final io.papermc.paper.configuration.WorldConfiguration.StructureSeeds conf // Spigot // Paper
     ) {
         List<Holder<StructureSet>> structures = structureOverrides.filter(structureSet -> hasBiomesForStructureSet(structureSet.value(), biomeSource)).toList();
-        return new ChunkGeneratorStructureState(randomState, biomeSource, levelSeed, 0L, structures);
+        return new ChunkGeneratorStructureState(randomState, biomeSource, levelSeed, 0L, ChunkGeneratorStructureState.injectSpigot(structures, conf), conf); // Spigot // Paper
     }
 
     public static ChunkGeneratorStructureState createForNormal(
-        final RandomState randomState, final long levelSeed, final BiomeSource biomeSource, final HolderLookup<StructureSet> allStructures
+        final RandomState randomState, final long levelSeed, final BiomeSource biomeSource, final HolderLookup<StructureSet> allStructures, final io.papermc.paper.configuration.WorldConfiguration.StructureSeeds conf // Spigot // Paper
     ) {
         List<Holder<StructureSet>> structures = allStructures.listElements()
             .filter(structureSet -> hasBiomesForStructureSet(structureSet.value(), biomeSource))
             .collect(Collectors.toUnmodifiableList());
-        return new ChunkGeneratorStructureState(randomState, biomeSource, levelSeed, levelSeed, structures);
-    }
+        return new ChunkGeneratorStructureState(randomState, biomeSource, levelSeed, levelSeed, ChunkGeneratorStructureState.injectSpigot(structures, conf), conf); // Spigot
+    }
+    // Paper start - Add missing structure set seed configs; horrible hack because spigot creates a ton of direct Holders which lose track of the identifying key
+    public static final class KeyedRandomSpreadStructurePlacement extends net.minecraft.world.level.levelgen.structure.placement.RandomSpreadStructurePlacement {
+        public final net.minecraft.resources.ResourceKey<StructureSet> key;
+        public KeyedRandomSpreadStructurePlacement(net.minecraft.resources.ResourceKey<StructureSet> key, net.minecraft.core.Vec3i locateOffset, FrequencyReductionMethod frequencyReductionMethod, float frequency, int salt, java.util.Optional<StructurePlacement.ExclusionZone> exclusionZone, int spacing, int separation, net.minecraft.world.level.levelgen.structure.placement.RandomSpreadType spreadType) {
+            super(locateOffset, frequencyReductionMethod, frequency, salt, exclusionZone, spacing, separation, spreadType);
+            this.key = key;
+        }
+    }
+    // Paper end - Add missing structure set seed configs
+
+    // Spigot start
+    // Paper start - Add missing structure set seed configs
+    private static List<Holder<StructureSet>> injectSpigot(final List<Holder<StructureSet>> list, final io.papermc.paper.configuration.WorldConfiguration.StructureSeeds conf) {
+        return list.stream().map((holder) -> {
+            net.minecraft.world.level.levelgen.structure.StructureSet structureset = holder.value();
+            final net.minecraft.core.Holder<net.minecraft.world.level.levelgen.structure.StructureSet> newHolder; // Paper - Add missing structure set seed configs
+            if (structureset.placement() instanceof net.minecraft.world.level.levelgen.structure.placement.RandomSpreadStructurePlacement randomConfig) { // Paper - Add missing structure set seed configs; check namespace cause datapacks could add structure sets with the same path
+                int seed = conf.structures.getOrDefault(holder, randomConfig.salt);
+                structureset = new net.minecraft.world.level.levelgen.structure.StructureSet(structureset.structures(), new net.minecraft.world.level.chunk.ChunkGeneratorStructureState.KeyedRandomSpreadStructurePlacement(holder.unwrapKey().orElseThrow(), randomConfig.locateOffset, randomConfig.frequencyReductionMethod, randomConfig.frequency, seed, randomConfig.exclusionZone, randomConfig.spacing(), randomConfig.separation(), randomConfig.spreadType()));
+                newHolder = net.minecraft.core.Holder.direct(structureset); // I really wish we didn't have to do this here // Maybe make salt non-final and replace with just: structureset.placement().salt = seed;
+            } else {
+                newHolder = holder;
+            }
+            return newHolder;
+            // Paper end - Add missing structure set seed configs
+        }).collect(Collectors.toUnmodifiableList());
+    }
+    // Spigot end
 
     private static boolean hasBiomesForStructureSet(final StructureSet structureSet, final BiomeSource biomeSource) {
         Stream<Holder<Biome>> structureBiomes = structureSet.structures().stream().flatMap(entry -> {
@@ -71,13 +_,14 @@
         final BiomeSource biomeSource,
         final long levelSeed,
         final long concentricRingsSeed,
-        final List<Holder<StructureSet>> possibleStructureSets
+        final List<Holder<StructureSet>> possibleStructureSets, final io.papermc.paper.configuration.WorldConfiguration.StructureSeeds conf // Paper - Add missing structure set seed configs
     ) {
         this.randomState = randomState;
         this.levelSeed = levelSeed;
         this.biomeSource = biomeSource;
         this.concentricRingsSeed = concentricRingsSeed;
         this.possibleStructureSets = possibleStructureSets;
+        this.conf = conf; // Paper - Add missing structure set seed configs
     }
 
     public List<Holder<StructureSet>> possibleStructureSets() {
@@ -116,7 +_,13 @@
         int spread = placement.spread();
         HolderSet<Biome> preferredBiomes = placement.preferredBiomes();
         RandomSource random = RandomSource.create();
+        // Paper start - Add missing structure set seed configs
+        if (structureSet.is(net.minecraft.world.level.levelgen.structure.BuiltinStructureSets.STRONGHOLDS)) {
+            random.setSeed(this.conf.structures.getOrDefault(structureSet, Math.toIntExact(this.concentricRingsSeed))); // FIXME: Maybe make all the structure seeds longs instead?
+        } else {
+        // Paper end - Add missing structure set seed configs
         random.setSeed(this.concentricRingsSeed);
+        } // Paper - Add missing structure set seed configs
         double angle = random.nextDouble() * Math.PI * 2.0;
         int positionInCircle = 0;
         int circle = 0;
@@ -193,7 +_,7 @@
 
         for (int testX = sourceX - range; testX <= sourceX + range; testX++) {
             for (int testZ = sourceZ - range; testZ <= sourceZ + range; testZ++) {
-                if (placement.isStructureChunk(this, testX, testZ)) {
+                if (placement.isStructureChunk(this, testX, testZ, structureSet)) { // Paper - Add missing structure set seed configs
                     return true;
                 }
             }
